framework/Optimizers/OptimizerBase.py

```diff
@@ -0,0 +1,515 @@
+    ready = self.localStillReady(ready)
+    return ready
+
+  def _incrementCounter(self):
+    """
+      Increments counter and sets up prefix.
+      @ In, None
+      @ Out, None
+    """
+    #since we are creating the input for the next run we increase the counter and global counter
+    self.counter['mdlEval'] += 1
+    self.inputInfo['prefix'] = str(self.counter['mdlEval'])
+
+  def updateVariableHistory(self,data,traj=0):
+    """
+      Stores new historical points into the optimization history.
+      @ In, data, dict, new input space entries as {var:#, var:#}
+      @ In, traj, int, integer label of the current trajectory of interest
+      @ Out, None
+    """
+    self.optVarsHist[traj][self.counter['varsUpdate'][traj]] = copy.deepcopy(data)
+
+  @abc.abstractmethod
+  def checkConvergence(self):
+    """
+      Method to check whether the convergence criteria has been met.
+      @ In, none,
+      @ Out, convergence, bool, variable indicating whether the convergence criteria has been met.
+    """
+
+  def clearCurrentOptimizationEffort(self):
+    """
+      Used to inform the subclass optimization effor that it needs to forget whatever opt data it is using
+      for the current point (for example, gradient determination points) so that we can start new.
+      @ In, None
+      @ Out, None
+    """
+    # README: this method is necessary because the base class optimizer doesn't know what needs to be reset in the
+    #         subclass, but the subclass doesn't know when it needs to call this method.
+    pass
+
+  def writeToSolutionExport(self, traj=0):
+    """
+      Standardizes how the solution export is written to.
+      Uses data from "recentOptHist" and other counters to fill in values.
+      @ In, traj, int, the trajectory for which an entry is being written
+      @ Out, None
+    """
+    pass
+
+  def _checkModelFinish(self, traj=0, updateKey=0, evalID='v'):
+    """
+      Determines if the Model has finished running an input and returned the output
+      @ In, traj, int, traj on which the input is being checked
+      @ In, updateKey, int, the id of variable update on which the input is being checked
+      @ In, evalID, int or string, indicating the id of the perturbation (int) or its a variable update (string 'v')
+      @ Out, _checkModelFinish, tuple(bool, int), indicating whether the Model has finished the evaluation over input
+             identified by traj+updateKey+evalID, the index of the location of the input in dataobject
+    """
+    if len(self.mdlEvalHist) == 0:
+      return (False,-1)
+    lookFor = '{}_{}_{}'.format(traj,updateKey,evalID)
+    index,match = self.mdlEvalHist.realization(matchDict = {'prefix':lookFor})
+    # if no match, return False
+    if match is None:
+      return False,-1
+    # otherwise, return index of match
+    return True, index
+
+  def _createEvaluationIdentifier(self, trajID=0, iterID=0, evalType='v'):
```

Comment:
Ah, and here is that call! Is this generic enough for arbitrary Optimizers, and then specializations can be done for specific ones, or should this be a virtual method that subclasses must overwrite?